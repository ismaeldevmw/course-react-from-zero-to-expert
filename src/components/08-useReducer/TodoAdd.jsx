import React, { useRef } from 'react'

export const TodoAdd = ({ handleAddTask }) => {

    const inputRef = useRef();

    const handleSubmit = (e) => {

        e.preventDefault();

        if (inputRef.current.value.trim().length <= 1) return;

        const newTask = {
            id: new Date().getTime(),
            description: inputRef.current.value,
            done: false
        };

        handleAddTask( newTask );
        inputRef.current.value = '';

    }

    return (
        <>
            <h4>Add task</h4>
            <hr/>

            <form onSubmit={ handleSubmit } className="form-inline">
                <div className="form-group mx-sm-3 mb-2">
                    <input 
                        type="text"
                        name="description"
                        placeholder="Add a task to do..."
                        autoComplete="off"
                        className="form-control form-control-sm"
                        ref={ inputRef }
                    />
                </div>
                <button
                    type="submit" 
                    className="btn btn-sm btn-outline-primary mb-2"
                >
                    Add
                </button>
            </form>
        </>
    )
}
