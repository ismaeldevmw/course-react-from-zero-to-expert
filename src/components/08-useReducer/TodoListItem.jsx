import React from 'react'

export const TodoListItem = ({ task, index, handleToggle, handleDelete }) => {
    return (
        <li
            key={ task.id }
            className="list-group-item d-flex justify-content-between align-items-center"
        >
            <span
                className={ `${ task.done && 'complete' }` } 
                onClick={ () => handleToggle( task.id ) } 
            >
                { index + 1}. { task.description }
            </span>
            <button onClick={ () => handleDelete(task.id) } className="btn btn-sm btn-danger">Delete</button>
        </li>
    )
}
