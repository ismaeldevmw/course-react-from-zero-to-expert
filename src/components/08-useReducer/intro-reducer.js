const initialState = [{
    id: 1,
    todo: 'Buy milk',
    done: false
}];

const todoReducer = ( state = initialState, action ) => {
    
    if ( action?.type === 'add' ) {
        return [ ...state, action.payload ];
    }

    return state;
}

let todos = todoReducer();


const newTodo = {
    id: 2,
    todo: 'But eggs',
    done: false
}


const addElementAction = {
    type: 'add',
    payload: newTodo
}


todos = todoReducer( todos, addElementAction );

console.log(todos);