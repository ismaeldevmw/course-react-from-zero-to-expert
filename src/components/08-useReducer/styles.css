body {
    padding: 3rem;
}

.complete {
    text-decoration: line-through;
}