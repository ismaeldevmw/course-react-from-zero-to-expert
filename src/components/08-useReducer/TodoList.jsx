import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ tasks = [], handleToggle, handleDelete }) => {
    return (
        <ul className="list-group">
            {
                tasks.map( (task, i) => (
                    <TodoListItem
                        key={ task.id }
                        task={ task } 
                        index={ i } 
                        handleToggle={ handleToggle }
                        handleDelete={ handleDelete } 
                    />
                ))
            }
        </ul>
    )
}
