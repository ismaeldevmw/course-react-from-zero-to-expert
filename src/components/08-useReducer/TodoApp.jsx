import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';

import './styles.css';

const init = () => {

    return JSON.parse(localStorage.getItem('tasks')) || []
    // return [{
    //     id: new Date().getTime(),
    //     description: 'Learn React',
    //     done: false
    // }];

}

export const TodoApp = () => {

    const [ taskToDo, dispatch ] = useReducer(todoReducer, [], init);

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    useEffect( () => {
        localStorage.setItem('tasks', JSON.stringify( taskToDo ));
    }, [ taskToDo ])

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) return;
        
        const newTask = {
            id: new Date().getTime(),
            description: description,
            done: false
        };

        const action = {
            type: 'add',
            payload: newTask
        }

        dispatch( action);
        reset();
    }

    return (
        <div>
            <h1>TodoApp ({ taskToDo.length })</h1>
            <hr/>

            <div className="row">
                <div className="col-7">
                    <ul className="list-group">
                        {
                            taskToDo.map( (task, i) => (
                                <li
                                    key={ task.id }
                                    className="list-group-item d-flex justify-content-between align-items-center"
                                >
                                    { i + 1}. { task.description }
                                    <button className="btn btn-sm btn-danger">Delete</button>
                                </li>
                            ))
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Add task</h4>
                    <hr/>

                    <form onSubmit={ handleSubmit } className="form-inline">
                        <div className="form-group mx-sm-3 mb-2">
                            <input 
                                type="text"
                                name="description"
                                placeholder="Add a task to do..."
                                autoComplete="off"
                                className="form-control form-control-sm"
                                value={ description }
                                onChange={ handleInputChange }
                            />
                        </div>
                        <button
                            type="submit" 
                            className="btn btn-sm btn-outline-primary mb-2"
                        >
                            Add
                        </button>
                    </form>
                </div>
            </div>
        </div>
    )
}
