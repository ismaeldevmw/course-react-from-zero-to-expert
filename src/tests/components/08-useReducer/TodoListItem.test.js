import React from 'react';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';
import { demoTasks } from "../../fixtures/demoTasks";

describe('Tests in <TodoListItem />', () => {

    const handleToggle = jest.fn();
    const handleDelete = jest.fn();
    
    const wrapper = shallow(
        <TodoListItem 
            task={ demoTasks[0] }
            index={ 0 }
            handleToggle={ handleToggle }
            handleDelete={ handleDelete }
        />
    );
    
    test('should display correctly', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('should call the handleDelete function', () => {
        
        wrapper.find('button').simulate('click');
        expect( handleDelete ).toHaveBeenCalledWith( demoTasks[0].id );

    });

    test('should call the handleToggle function', () => {
        
        wrapper.find('span').simulate('click');
        expect( handleToggle ).toHaveBeenCalledWith( demoTasks[0].id );

    });

    test('should show the text correctly', () => {
        
        const p = wrapper.find('span');
        expect( p.text().trim() ).toBe(`1. ${ demoTasks[0].description }`)

    });

    test('should have the complete class if a task is done', () => {
        
        const task = demoTasks[0];
        task.done = true;
        
        const wrapper = shallow(
            <TodoListItem 
                task={ demoTasks[0] }
                index={ 0 }
            />
        );

        expect( wrapper.find('span').hasClass('complete') ).toBe(true);

    });

});