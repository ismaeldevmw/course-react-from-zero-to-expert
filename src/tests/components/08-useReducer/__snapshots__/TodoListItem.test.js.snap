// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoListItem /> should display correctly 1`] = `
<li
  className="list-group-item d-flex justify-content-between align-items-center"
  key="1"
>
  <span
    className="false"
    onClick={[Function]}
  >
    1
    . 
    Learn React
  </span>
  <button
    className="btn btn-sm btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
